File: 140-word-break-ii/word-break-ii.java
class Solution {
    List<String> ll = new ArrayList<>();
    public List<String> wordBreak(String s, List<String> wordDict) {
        HashSet<String> set = new HashSet<>(wordDict);
        word(s, set, 0, 0, "");
        return ll;
    }
    public void word(String s, HashSet<String> set, int i, int j, String ans) {
        if(i == s.length() && j == s.length()) {
            ll.add(ans);
            return;
        }
        if(j >= s.length()) {
            return;
        }
        String s1 = s.substring(i, j+1);
        if(set.contains(s1)) {
            if(ans.length() == 0) {
                word(s, set, j+1, j+1, ans+s1);
            }
            else {
                word(s, set, j+1, j+1, ans+" "+s1);
            }
            word(s, set, i, j+1, ans);
        }
        else {
            word(s, set, i, j+1, ans);
        }
    }
}

File: 140-word-break-ii/word-break-ii.cpp
class Solution {
public:
    vector<string> ans;
    unordered_set<string> set;
    vector<string> wordBreak(string s, vector<string>& wordDict) {
        
        for(string s1: wordDict) {
            set.insert(s1);
        }

        word(s, 0, 0, "");

        return ans;
    }

    void word(string& s, int i, int j, string s1) {
        if(j == s.length()) {
            if(i == s.length()) {
                ans.push_back(s1);
            }
            return;
        }

        string s2 = s.substr(i, j-i+1);

        if(set.count(s2)) {

            if(s1.length() == 0) {
                word(s, j+1, j+1, s2);
            }
            else {
                word(s, j+1, j+1, s1+" "+s2);
            }
        }

        word(s, i, j+1, s1);
    }
};